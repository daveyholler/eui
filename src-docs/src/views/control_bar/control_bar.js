import React, { Component } from 'react';

import {
  EuiButton,
  EuiControlBar,
  EuiLink,
  EuiText,
} from '../../../../src/components';

export default class extends Component {
  constructor(props) {
    super(props);
    this.state = {
      contentIsVisible: false,
      tabContent: '',
    };
  }

  toggle() {
    this.setState({
      contentIsVisible: !this.state.contentIsVisible,
    });
  }

  soundTheAlarms = () => {
    alert('You clicked a button!');
  };

  closeTheHatch = () => {
    this.setState({
      contentIsVisible: false,
    });
  };

  tabOne = () => {
    this.setState({
      contentIsVisible: true,
      tabContent:
        "Oceanic Airlines Flight 815 was a scheduled flight from Sydney, Australia to Los Angeles, California, United States, on a Boeing 777-200ER. On September 22, 2004 at 4:16 P.M., the airliner, carrying 324 passengers, deviated from its original course and disappeared over the Pacific Ocean. This is the central moment in the series that kicked off its plotline, and marked the chronological beginning of the main characters' adventures on the Island.",
    });
  };

  tabTwo = () => {
    this.setState({
      contentIsVisible: true,
      tabContent:
        'The Others, referred to by the DHARMA Initiative as the Hostiles or the Natives, and also by the tail section survivors of Oceanic Flight 815 as Them, are a group of people living on the Island who were followers of Jacob, intermediated by Richard Alpert. Jacob never showed himself to his people, and they took orders from a succession of leaders including Eloise Hawking, Charles Widmore, Benjamin Linus, and briefly, John Locke.',
    });
  };
  render() {
    const textLink = (
      <EuiLink href="https://www.youtube.com/watch?v=dQw4w9WgXcQ">
        src/component/roller.tsx
      </EuiLink>
    );
    const controls = [
      {
        id: 'sound_the_alarm',
        label: 'Sound the Alarm',
        controlType: 'button',
        onClick: this.soundTheAlarms,
        color: 'warning',
      },
      {
        id: 'close_the_hatch',
        label: 'Close the Hatch',
        controlType: 'button',
        onClick: this.closeTheHatch,
        classNames: 'customClassName',
        color: 'primary',
      },
      {
        id: 'flight_815',
        label: 'Flight 815',
        controlType: 'tab',
        onClick: this.tabOne,
      },
      {
        id: 'the_others',
        label: 'The Others',
        controlType: 'tab',
        onClick: this.tabTwo,
      },
      {
        id: 'spacer_1',
        controlType: 'spacer',
      },
      {
        id: 'set_the_timer',
        label: 'Set the Timer',
        controlType: 'icon',
        iconType: 'clock',
        onClick: this.closeTheHatch,
      },
      {
        id: 'some_text',
        label: textLink,
        controlType: 'text',
        onClick: null,
      },
    ];

    return (
      <div
        style={{
          width: '100%',
          height: '400px',
          position: 'relative',
        }}>
        <EuiButton onClick={this.toggle.bind(this)}>Toggle</EuiButton>
        <EuiControlBar
          controls={controls}
          showContent={this.state.contentIsVisible}>
          <div style={{ padding: '1rem' }}>
            {this.state.tabContent !== '' ? (
              <EuiText>{this.state.tabContent}</EuiText>
            ) : (
              <p>Look at me</p>
            )}
          </div>
        </EuiControlBar>
      </div>
    );
  }
}
