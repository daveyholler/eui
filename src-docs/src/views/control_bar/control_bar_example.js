import React from 'react';

import {
  EuiBasicTable,
  EuiCode,
  EuiControlBar,
} from '../../../../src/components';

import { renderToHtml } from '../../services';

import { GuideSectionTypes } from '../../components';

import ControlBar from './control_bar';
import { ControlBarWithTabs } from './tabs';

const controlBarSource = require('!!raw-loader!./control_bar');
const controlBarHtml = renderToHtml(ControlBar);
const controlBarSnippet = '<EuiControlBar controls={items}/>';

const tabsBarSource = require('!!raw-loader!./tabs');
const tabsBarHtml = renderToHtml(ControlBarWithTabs);
const tabsBarSnippet = '<EuiControlBar controls={items} size="m"/>';

const buttonPropsTable = [
  {
    prop: 'id',
    type: 'string',
    description: 'Provide a unique id for the element',
  },
  {
    prop: 'color (optional)',
    type: 'props of EuiButton["color"]',
    description: 'Set the button color. Defaults to "ghost."',
  },
  {
    prop: 'label',
    type: 'string',
    description: 'Displays the button label',
  },
  {
    prop: 'classNames (optional)',
    type: 'string',
    description: 'Optionally pass additional classes.',
  },
  {
    prop: 'onClick',
    type: 'function',
    description: 'Handle the click event',
  },
];

const tabPropsTable = [
  {
    prop: 'id',
    type: 'string',
    description: 'Provide a unique id for the element',
  },
  {
    prop: 'label',
    type: 'string',
    description: 'Displays the Tab label',
  },
  {
    prop: 'onClick',
    type: 'function',
    description:
      'Handle the tab click. Most likely this will be used to set ' +
      'something like the active tab on your state.',
  },
];

const breadcrumbsPropsTable = [
  {
    prop: 'id',
    type: 'string',
    description: 'Provide a unique id for the element',
  },
  {
    prop: 'responsive (optional)',
    type: 'boolean',
    description: 'Hides left most breadcrumbs as window gets smaller',
  },
  {
    prop: 'truncate (optional)',
    type: 'boolean',
    description:
      'Forces all breadcrumbs to single line and' +
      'truncates each breadcrumb to a particular width, except for the last item',
  },
  {
    prop: 'max (optional)',
    type: 'number',
    description:
      'Condenses the inner items past the maximum set into a single ellipses item',
  },
  {
    prop: 'breadcrumbs',
    type: 'array',
    description: 'An array of items for the breadcrumbs',
  },
];

const textPropsTable = [
  {
    prop: 'id',
    type: 'string',
    description: 'Provide a unique id for the element',
  },
  {
    prop: 'label',
    type: 'string',
    description: 'Displays the Text label',
  },
  {
    prop: 'color (optional)',
    type: 'props of EuiText["color"]',
    description: 'Sets the color of the text. Defaults to "ghost."',
  },
  {
    prop: 'onClick (optional)',
    type: 'function',
    description: 'Handle the click event',
  },
];

const iconPropsTable = [
  {
    prop: 'id',
    type: 'string',
    description: 'Provide a unique id for the element.',
  },
  {
    prop: 'iconType',
    type: 'string',
    description: 'Which icon to render',
  },
  {
    prop: 'label',
    type: 'string',
    description: 'For accessibility',
  },
  {
    prop: 'classNames (optional)',
    type: 'string',
    description: 'Optionally pass additional classes.',
  },
  {
    prop: 'color (optional)',
    type: 'props of EuiButtonIcon["color"]',
    description: 'Sets the color of the icon. Defaults to "ghost."',
  },
  {
    prop: 'onClick (optional)',
    type: 'function',
    description: 'Handle the click event',
  },
];

const tableColumns = [
  {
    field: 'prop',
    name: 'Prop',
    sortable: false,
    'data-test-subj': 'propCell',
  },
  {
    field: 'type',
    name: 'Type',
    sortable: false,
    'data-test-subj': 'typeCell',
  },
  {
    field: 'description',
    name: 'Description',
    sortable: false,
    'data-test-subj': 'descriptionCell',
  },
];

export const ControlBarExample = {
  title: 'Control Bar',
  sections: [
    {
      source: [
        {
          type: GuideSectionTypes.JS,
          code: controlBarSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: controlBarHtml,
        },
      ],
      text: (
        <div>
          <p>
            <EuiCode>ControlBar</EuiCode> is a bottom positioned container and
            content well intended to provide additional view controls and
            actions.
          </p>
          <h2>Mobile Usage</h2>
          The <EuiCode>ControlBar</EuiCode> is responsive in the sense that it
          utilizes flexbox. However, it makes no attempts to reorganize the
          controls you provide. By default the <EuiCode>ControlBar</EuiCode> is
          hidden on mobile devices, but this can be overridden with the
          <EuiCode>showOnMobile</EuiCode> prop. You&apos;ll need to take smaller
          screens into consideration when choosing to display on smaller
          screens.
        </div>
      ),
      props: { EuiControlBar },
      snippet: controlBarSnippet,
      demo: <ControlBar />,
    },
    {
      title: 'Using Tabs',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: tabsBarSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: tabsBarHtml,
        },
      ],
      text: (
        <div>
          <p>
            This example deomnstrates the use of tabs and uses the size
            <EuiCode>size=&quot;m&quot;</EuiCode>.
          </p>
        </div>
      ),
      props: { EuiControlBar },
      snippet: tabsBarSnippet,
      demo: <ControlBarWithTabs />,
    },
    {
      title: 'ControlTypes',
      text: (
        <div>
          <p>
            The <EuiCode>ControlBar</EuiCode> accepts an array of
            <EuiCode>controlTypes</EuiCode> that can be arranged virtually any
            way you&apos;d like by using any of following
            <EuiCode>controlTypes</EuiCode>. The controlTypes will be ordered
            the same way they are in the array you provide.
          </p>
          <h3>Button Control</h3>
          <EuiBasicTable items={buttonPropsTable} columns={tableColumns} />
          <h3>Tab Control</h3>
          <EuiBasicTable items={tabPropsTable} columns={tableColumns} />
          <h3>Breadcrumbs Control</h3>
          <p>
            View the documentation on <EuiCode>EuiBreadcrumbs</EuiCode> for
            additional information on their usage.
          </p>
          <EuiBasicTable items={breadcrumbsPropsTable} columns={tableColumns} />
          <h3>Text Control</h3>
          <EuiBasicTable items={textPropsTable} columns={tableColumns} />
          <h3>Icon Control</h3>
          <EuiBasicTable items={iconPropsTable} columns={tableColumns} />
          <h3>Spacer & Divider Control</h3>
          <h4>Spacers</h4>
          <p>
            Spacers can be used to provide horizontal spaces between your
            <EuiCode>controlTypes</EuiCode>. Spacers do not need{' '}
            <EuiCode>ids</EuiCode>.
          </p>
          <h4>Dividers</h4>
          <p>
            Dividers provide <EuiCode>1px</EuiCode> wide colored breaks between
            your <EuiCode>controlTypes</EuiCode>. Useful when additional visual
            separation is needed. Dividers do not need <EuiCode>ids</EuiCode>.
          </p>
          <h3>Rendering content to the control bar drawer</h3>
          <p>
            Optional children of the <EuiCode>EuiControlBar</EuiCode> are
            rendered in the control bar drawer. You can toggle the visibility of
            the content with the <EuiCode>showContent</EuiCode> prop. When you
            want to display tab content, this is where you&apos;ll do it.
          </p>
        </div>
      ),
    },
  ],
};
