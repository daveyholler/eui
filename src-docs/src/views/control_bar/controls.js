import React from 'react';

import {
  EuiButton,
  EuiControlBar,
  EuiFlexGroup,
  EuiFlexItem,
  EuiText,
} from '../../../../src/components';
import { keyCodes } from '../../../../src/services';
import { EuiFocusTrap } from '../../../../src/components/focus_trap';

export class Controls extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      contentIsVisible: true,
      isFullScreen: true,
      tabContent: '',
    };
  }

  soundTheAlarms = () => {
    alert('You clicked a button!');
  };

  tabOne = () => {
    this.setState({
      contentIsVisible: true,
      tabContent: 'Tab content',
    });
  };

  toggle() {
    this.setState({
      contentIsVisible: !this.state.contentIsVisible,
    });
  }

  toggleFullScreen() {
    this.setState(prevState => ({
      isFullScreen: !prevState.isFullScreen,
      contentIsVisible: false,
    }));
  }

  onKeyDown = event => {
    if (event.keyCode === keyCodes.ESCAPE) {
      event.preventDefault();
      event.stopPropagation();
      this.setState({
        isFullScreen: false,
        contentIsVisible: false,
      });
    }
  };

  render() {
    const controls = [
      {
        controlType: 'button',
        id: 'controls_button',
        label: 'Button',
        onClick: this.soundTheAlarms,
      },
      {
        controlType: 'spacer',
      },
      {
        controlType: 'icon',
        id: 'controls_icon',
        iconType: 'flag',
      },
      {
        controlType: 'divider',
      },
      {
        controlType: 'icon',
        id: 'controls_icon_button',
        iconType: 'bell',
        onClick: this.soundTheAlarms,
        color: 'primary',
      },
      {
        controlType: 'divider',
      },
      {
        controlType: 'text',
        id: 'controls_text',
        label: 'Some text',
      },
      {
        controlType: 'divider',
      },
      {
        controlType: 'tab',
        id: 'controls_tab',
        label: 'Tab',
        onClick: this.tabOne,
      },
      {
        controlType: 'spacer',
      },
      {
        controlType: 'breadcrumbs',
        id: 'controls_breadcrumbs',
        breadcrumbs: [
          {
            text: 'Breadcrumbs',
          },
          {
            text: 'Item',
          },
        ],
      },
    ];

    let fullScreenDisplay;

    if (this.state.isFullScreen) {
      fullScreenDisplay = (
        <EuiFocusTrap>
          <div className="guideDemo__pageOverlay" onKeyDown={this.onKeyDown}>
            <EuiFlexGroup>
              <EuiButton onClick={this.toggle.bind(this)}>
                Toggle content drawer
              </EuiButton>
              <EuiFlexItem grow />
              <EuiButton onClick={this.toggleFullScreen.bind(this)}>
                Close full screen
              </EuiButton>
            </EuiFlexGroup>
            <EuiControlBar
              controls={controls}
              size="m"
              showContent={this.state.contentIsVisible}
              showOnMobile>
              <div style={{ padding: '1rem' }}>
                {this.state.tabContent !== '' ? (
                  <EuiText>{this.state.tabContent}</EuiText>
                ) : (
                  <p>Look at me</p>
                )}
              </div>
            </EuiControlBar>
          </div>
        </EuiFocusTrap>
      );
    }

    return (
      <div>
        <EuiButton onClick={this.toggleFullScreen.bind(this)}>
          View in full screen
        </EuiButton>
        {fullScreenDisplay}
      </div>
    );
  }
}
