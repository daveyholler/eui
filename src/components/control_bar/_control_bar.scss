.euiControlBar {
  background: lighten($euiColorInk, 5%);
  width: calc(100% - 3rem);
  color: $controlBarText;
  display: flex;
  flex-direction: column;
  position: fixed;
  left: 3rem;
  box-shadow: inset 0 $euiSizeXXL 0 $controlBarBackground, inset 0 600rem 0 $euiColorLightestShade;
  z-index: $euiZNavigation;
  transform: translateY(0);

  &.euiControlBar--open {
    animation-duration: $euiAnimSpeedNormal;
    animation-timing-function: $euiAnimSlightResistance;
    animation-fill-mode: forwards;
  }

  &.euiControlBar--open.euiControlBar--large {
    animation-name: openPanelLarge;
  }

  &.euiControlBar--open.euiControlBar--medium {
    animation-name: openPanelMedium;
  }

  &.euiControlBar--open.euiControlBar--small {
    animation-name: openPanelMedium;
  }

  &.euiControlBar--large {
    height: 90%;
    bottom: calc((90% - #{$euiSizeXXL}) * -1);
  }

  &.euiControlBar--medium {
    height: 50%;
    bottom: calc((50% - #{$euiSizeXXL}) * -1);
  }

  &.euiControlBar--small {
    height: 25%;
    bottom: calc((25% - #{$euiSizeXXL}) * -1);
  }
}

.euiControlBar__controls {
  height: $euiSizeXXL;
  width: 100%;
  display: flex;
  align-items: center;
}

.euiControlBar__content {
  width: 100%;
  height: calc(100% - #{$euiSizeXXL});
  background-color: $euiColorLightestShade;
  animation-name: showContent;
  animation-duration: $euiAnimSpeedSlow;
  animation-iteration-count: 1;
  animation-timing-function: $euiAnimSlightResistance;
  color: $euiColorDarkestShade;
  overflow-y: scroll;
}

@keyframes openPanelLarge {
  0% {
    transform: translateY(0);
  }

  100% {
    transform: translateY(-90%);
  }
}

@keyframes openPanelMedium {
  0% {
    transform: translateY(0);
  }

  100% {
    transform: translateY(-50%);
  }
}

@keyframes openPanelSmall {
  0% {
    transform: translateY(0);
  }

  100% {
    transform: translateY(-25%);
  }
}

@keyframes showContent {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}

.euiControlBar__euiButtonIcon {
  min-width: $euiSizeXXL;
}

.euiControlBar__euiButton {
  border-radius: 0;
  border-left: none;
  border-top: none;
  border-bottom: none;

  &:enabled:hover {
    transform: translateY(0);
    box-shadow: none;
  }

  &:last-child {
    border-right: none;
  }

}

.euiControlBar__spacer {
  display: flex;
  flex-grow: 1;
  height: 100%;
}

.euiControlBar__divider {
  height: 100%;
  width: 1px;
  background-color: $controlBarBorderColor;
}

.euiControlBar__euiText {
  padding: 0 $euiSizeS;
  color: $controlBarText;
}

.euiControlBar__tab {
  padding: 0 $euiSize;
  text-align: center;
  box-sizing: border-box;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  color: $controlBarText;

  &:hover {
    text-decoration: underline;
    cursor: pointer;
  }

  &.euiControlBar__tab--active {
    background-color: $euiColorLightestShade;
    box-shadow: inset 0 4px 0 makeHighContrastColor($euiColorPrimary, $euiColorLightestShade);
    color: makeHighContrastColor($euiColorPrimary, $euiColorLightestShade);
  }
}

.euiControlBar__euiButton + .euiControlBar__tab--active {
  box-shadow: inset 0 4px 0 makeHighContrastColor($euiColorPrimary, $euiColorLightestShade), -1px 0 0 $controlBarBackground;
}
