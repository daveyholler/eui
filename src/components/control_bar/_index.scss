@import '../button/variables';
@import 'variables';
@import 'control_bar';
