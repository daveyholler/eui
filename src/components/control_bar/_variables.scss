$controlBarBackground: lightOrDarkTheme($euiColorDarkestShade, $euiColorInk);
$controlBarText: lighten(makeHighContrastColor($euiColorLightestShade, $controlBarBackground), 20%);
$controlBarBorderColor: rgba(255, 255, 255, .2);

@each $colorName, $colorValue in $buttonTypes {
  .euiControlBar__controls {
    .euiText--#{$colorName} {
      color: makeHighContrastColor($colorValue, $controlBarBackground);
    }

    .euiLink.euiLink--#{$colorName} {
      color: makeHighContrastColor($colorValue, $controlBarBackground);

      &:hover {
        color: tintOrShade($colorValue, 30%, 30%);
      }
    }

    .euiButton--#{$colorName}:enabled {
      color: makeHighContrastColor($colorValue, $controlBarBackground);
      border-color: $controlBarBorderColor;
      box-shadow: none;
    }

    .euiIcon--#{$colorName} {
      color: makeHighContrastColor($colorValue, $controlBarBackground);
    }
  }
}
