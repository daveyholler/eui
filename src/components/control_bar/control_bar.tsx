import React, { Component, HTMLAttributes, ButtonHTMLAttributes } from 'react';
import classnames from 'classnames';
import { CommonProps, PropsOf, ExclusiveUnion } from '../common';
// @ts-ignore-next-line
import { EuiBreadcrumbs } from '../breadcrumbs';
// @ts-ignore-next-line
import { EuiButton, EuiButtonIcon } from '../button';
import { EuiPortal } from '../portal';
import { EuiText } from '../text';

type ButtonControl = ButtonHTMLAttributes<HTMLButtonElement> & {
  controlType: 'button';
  id: string;
  color?: PropsOf<typeof EuiButton>['color'];
  label: string;
  classNames?: string;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
};

type TabControl = ButtonHTMLAttributes<HTMLButtonElement> & {
  controlType: 'tab';
  id: string;
  label: string;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
};

interface Breadcrumb {
  text: string;
  href?: string;
  onClick?: React.MouseEventHandler<HTMLDivElement>;
  truncate?: boolean;
}

type BreadcrumbControl = HTMLAttributes<HTMLDivElement> & {
  controlType: 'breadcrumbs';
  id: string;
  responsive?: boolean;
  truncate?: boolean;
  max?: number;
  breadcrumbs: Breadcrumb[];
};

type TextControl = HTMLAttributes<HTMLDivElement> & {
  controlType: 'text';
  id: string;
  label: string;
  color?: PropsOf<typeof EuiText>['color'];
  onClick?: React.MouseEventHandler<HTMLDivElement>;
};

interface SpacerControl {
  controlType: 'spacer';
}

interface DivideControl {
  controlType: 'divider';
}

type IconControl = ButtonHTMLAttributes<HTMLButtonElement> & {
  controlType: 'icon';
  id: string;
  iconType: string;
  label: string;
  classNames?: string;
  color?: PropsOf<typeof EuiButtonIcon>['color'];
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
};

export type Control = ExclusiveUnion<
  ExclusiveUnion<
    ExclusiveUnion<
      ExclusiveUnion<
        ExclusiveUnion<
          ButtonControl,
          ExclusiveUnion<BreadcrumbControl, TabControl>
        >,
        TextControl
      >,
      IconControl
    >,
    DivideControl
  >,
  SpacerControl
>;

export type EuiControlBarProps = HTMLAttributes<HTMLDivElement> &
  CommonProps & {
    /**
     * Show or hide the content well with your custom content inside
     */
    showContent?: boolean;

    /**
     * An array of controls, actions, and layout spacers to display.
     * Accepts `'button' | 'tab' | 'breadcrumbs' | 'text' | 'icon' | 'spacer' | 'divider'`
     */
    controls: Control[];
    /**
     * The maximum height of the overlay (relative to the window). Default is 90%, Medium is 75%, Small is 50%;
     */
    size?: 's' | 'm' | 'l';
    /**
     * Set the offset from the left side of the screen to account for Kibana's left-hand navigation menu.
     */
    leftOffset?: 's' | 'l' | undefined;
  };

interface EuiControlBarState {
  selectedTab: string;
}

export class EuiControlBar extends Component<
  EuiControlBarProps,
  EuiControlBarState
> {
  state = {
    selectedTab: '',
  };

  render() {
    const {
      children,
      className,
      showContent,
      controls,
      size,
      leftOffset,
      ...rest
    } = this.props;

    const classes = classnames('euiControlBar', className, {
      'euiControlBar--open': showContent,
      'euiControlBar--large': size === 'l' || !size,
      'euiControlBar--medium': size === 'm',
      'euiControlBar--small': size === 's',
      'euiControlBar--navExpanded': leftOffset === 'l',
      'euiControlBar--navCollapsed': leftOffset === 's',
    });

    const tabClasses = classnames('euiControlBar__tab', {
      'euiControlBar__tab--active': showContent,
    });

    const handleTabClick = (
      control: TabControl,
      e: React.MouseEvent<HTMLButtonElement, MouseEvent>
    ) => {
      this.setState(
        {
          selectedTab: control.id,
        },
        () => {
          control.onClick(e);
        }
      );
    };

    const controlItem = (control: Control, index: number) => {
      switch (control.controlType) {
        case 'button': {
          const {
            controlType,
            id,
            color,
            label,
            classNames,
            onClick,
            ...rest
          } = control;
          return (
            <EuiButton
              key={id + index}
              onClick={onClick}
              className={classnames('euiControlBar__button', classNames)}
              color={color ? color : 'ghost'}
              {...rest}>
              <EuiText size="s">{label}</EuiText>
            </EuiButton>
          );
        }
        case 'icon': {
          const {
            controlType,
            id,
            iconType,
            label,
            classNames,
            color,
            onClick,
            ...rest
          } = control;
          return (
            <EuiButtonIcon
              key={id + index}
              iconType={iconType}
              data-test-subj={label}
              aria-label={label}
              onClick={onClick}
              className={classnames('euiControlBar__buttonIcon', classNames)}
              color={color ? color : 'ghost'}
              {...rest}
            />
          );
        }
        case 'divider':
          return (
            <div
              key={control.controlType + index}
              className="euiControlBar__divider"
            />
          );
        case 'spacer':
          return (
            <div
              key={control.controlType + index}
              className="euiControlBar__spacer"
            />
          );
        case 'text': {
          const { controlType, id, label, color, onClick, ...rest } = control;
          return (
            <EuiText
              color={color ? color : 'ghost'}
              className="euiControlBar__euiText"
              key={id + index}
              size="s"
              {...rest}>
              {label}
            </EuiText>
          );
        }
        case 'tab': {
          const { controlType, id, label, onClick, ...rest } = control;
          return (
            <button
              key={id + index}
              className={`euiControlBar__tab ${
                id === this.state.selectedTab ? tabClasses : ''
              }`}
              data-test-subj={label}
              aria-label={`Control Bar - ${label}`}
              onClick={event => handleTabClick(control, event)}
              {...rest}>
              <EuiText size="s">{label}</EuiText>
            </button>
          );
        }
        case 'breadcrumbs': {
          const {
            controlType,
            id,
            responsive,
            truncate,
            max,
            breadcrumbs,
            ...rest
          } = control;
          return (
            <EuiBreadcrumbs
              key={control.id}
              breadcrumbs={control.breadcrumbs}
              responsive={control.responsive}
              truncate={control.truncate}
              max={control.max}
              {...rest}
            />
          );
        }
      }
    };

    return (
      <EuiPortal>
        <div className={classes} aria-label="Control Bar" {...rest}>
          <div className="euiControlBar__controls">
            {controls.map((control, index) => {
              return controlItem(control, index);
            })}
          </div>
          {this.props.showContent ? (
            <div className="euiControlBar__content">{children}</div>
          ) : null}
        </div>
      </EuiPortal>
    );
  }
}
