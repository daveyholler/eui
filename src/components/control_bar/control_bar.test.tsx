import React from 'react';
import { mount } from 'enzyme';
import { requiredProps } from '../../test/required_props';
import { EuiControlBar, Control } from './control_bar';

const handleClick = () => {
  console.log('You clicked');
};

const controls: Control[] = [
  {
    id: 'sound_the_alarm',
    label: 'Sound the Alarm',
    controlType: 'button',
    onClick: handleClick,
  },
  {
    id: 'close_the_hatch',
    label: 'Close the Hatch',
    controlType: 'button',
    onClick: handleClick,
  },
  {
    id: 'sample_icon',
    label: 'Sample Icon',
    controlType: 'icon',
    iconType: 'alert',
  },
];

describe('EuiControlBar', () => {
  test('is rendered', () => {
    const component = mount(
      <EuiControlBar controls={controls} {...requiredProps} />
    );

    expect(component).toMatchSnapshot();
  });
});
