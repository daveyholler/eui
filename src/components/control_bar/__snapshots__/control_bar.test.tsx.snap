// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiControlBar is rendered 1`] = `
<EuiControlBar
  aria-label="aria-label"
  className="testClass1 testClass2"
  controls={
    Array [
      Object {
        "controlType": "button",
        "id": "sound_the_alarm",
        "label": "Sound the Alarm",
        "onClick": [Function],
      },
      Object {
        "controlType": "button",
        "id": "close_the_hatch",
        "label": "Close the Hatch",
        "onClick": [Function],
      },
      Object {
        "controlType": "icon",
        "iconType": "alert",
        "id": "sample_icon",
        "label": "Sample Icon",
      },
    ]
  }
  data-test-subj="test subject string"
>
  <EuiPortal>
    <Portal
      containerInfo={
        <div>
          <div
            aria-label="aria-label"
            class="euiControlBar testClass1 testClass2 euiControlBar--large"
            data-test-subj="test subject string"
          >
            <div
              class="euiControlBar__controls"
            >
              <button
                aria-label="Control Bar - Sound the Alarm"
                class="euiButton euiButton--ghost euiControlBar__euiButton"
                data-test-subj="Sound the Alarm"
                type="button"
              >
                <span
                  class="euiButton__content"
                >
                  <span
                    class="euiButton__text"
                  >
                    <div
                      class="euiText euiText--small"
                    >
                      Sound the Alarm
                    </div>
                  </span>
                </span>
              </button>
              <button
                aria-label="Control Bar - Close the Hatch"
                class="euiButton euiButton--ghost euiControlBar__euiButton"
                data-test-subj="Close the Hatch"
                type="button"
              >
                <span
                  class="euiButton__content"
                >
                  <span
                    class="euiButton__text"
                  >
                    <div
                      class="euiText euiText--small"
                    >
                      Close the Hatch
                    </div>
                  </span>
                </span>
              </button>
              <button
                aria-label="Sample Icon"
                class="euiButtonIcon euiButtonIcon--ghost euiControlBar__euiButtonIcon"
                data-test-subj="Sample Icon"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                  focusable="false"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                />
              </button>
            </div>
          </div>
        </div>
      }
    >
      <div
        aria-label="aria-label"
        className="euiControlBar testClass1 testClass2 euiControlBar--large"
        data-test-subj="test subject string"
      >
        <div
          className="euiControlBar__controls"
        >
          <EuiButton
            aria-label="Control Bar - Sound the Alarm"
            className="euiControlBar__euiButton"
            color="ghost"
            data-test-subj="Sound the Alarm"
            key="sound_the_alarm0"
            onClick={[Function]}
          >
            <button
              aria-label="Control Bar - Sound the Alarm"
              className="euiButton euiButton--ghost euiControlBar__euiButton"
              data-test-subj="Sound the Alarm"
              onClick={[Function]}
              type="button"
            >
              <span
                className="euiButton__content"
              >
                <span
                  className="euiButton__text"
                >
                  <EuiText
                    size="s"
                  >
                    <div
                      className="euiText euiText--small"
                    >
                      Sound the Alarm
                    </div>
                  </EuiText>
                </span>
              </span>
            </button>
          </EuiButton>
          <EuiButton
            aria-label="Control Bar - Close the Hatch"
            className="euiControlBar__euiButton"
            color="ghost"
            data-test-subj="Close the Hatch"
            key="close_the_hatch1"
            onClick={[Function]}
          >
            <button
              aria-label="Control Bar - Close the Hatch"
              className="euiButton euiButton--ghost euiControlBar__euiButton"
              data-test-subj="Close the Hatch"
              onClick={[Function]}
              type="button"
            >
              <span
                className="euiButton__content"
              >
                <span
                  className="euiButton__text"
                >
                  <EuiText
                    size="s"
                  >
                    <div
                      className="euiText euiText--small"
                    >
                      Close the Hatch
                    </div>
                  </EuiText>
                </span>
              </span>
            </button>
          </EuiButton>
          <EuiButtonIcon
            aria-label="Sample Icon"
            className="euiControlBar__euiButtonIcon"
            color="ghost"
            data-test-subj="Sample Icon"
            iconType="alert"
            key="sample_icon2"
          >
            <button
              aria-label="Sample Icon"
              className="euiButtonIcon euiButtonIcon--ghost euiControlBar__euiButtonIcon"
              data-test-subj="Sample Icon"
              type="button"
            >
              <EuiIcon
                aria-hidden="true"
                className="euiButtonIcon__icon"
                size="m"
                type="alert"
              >
                <EuiIconEmpty
                  aria-hidden="true"
                  className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                  focusable="false"
                  style={null}
                >
                  <svg
                    aria-hidden="true"
                    className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                    focusable="false"
                    height={16}
                    style={null}
                    viewBox="0 0 16 16"
                    width={16}
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </EuiIconEmpty>
              </EuiIcon>
            </button>
          </EuiButtonIcon>
        </div>
      </div>
    </Portal>
  </EuiPortal>
</EuiControlBar>
`;
